src/analyses/base.ml

```diff
@@ -1632,14 +1631,69 @@ struct
         if M.tracing then M.tracel "inv" "Unhandled operator %a\n" d_binop op;
         a, b
     in
+    let inv_bin_float (a, b) fkind c op =
+      let meet_bin a' b'  = FD.meet a a', FD.meet b b' in
+      let meet_com oi = (* commutative *)
+        meet_bin (oi c b) (oi c a) in
+      let meet_non oi oo = (* non-commutative *)
+        meet_bin (oi c b) (oo a c) in
+      match op with
+      | PlusA  -> meet_com FD.sub
+      | Mult   ->
+        (* refine x by information about y, using x * y == c *)
+        let refine_by x y = (match FD.to_float y with
+            | None -> x
+            | Some _ -> FD.meet x (FD.div c y))
+        in
+        (refine_by a b, refine_by b a)
+      | MinusA -> meet_non FD.add FD.sub
+      | Div    ->
+        (* If b must be zero, we have must UB *)
+        meet_bin (FD.mul b c) (FD.div a c)
+      | Eq | Ne as op ->
+        let both x = x, x in
+        let m = FD.meet a b in
+        let result = FD.ne c (FD.of_const 0.) in
+        (match op, ID.to_bool(result) with
+         | Eq, Some true
+         | Ne, Some false -> both m (* def. equal: if they compare equal, both values must be from the meet *)
+         | Eq, Some false
+         | Ne, Some true -> (* def. unequal *)
+           (* M.debug ~category:Analyzer "Can't use uneqal information about float value in expression \"%a\"." d_plainexp exp; *)
+           a, b
+         | _, _ -> a, b
+        )
+      | Lt | Le | Ge | Gt as op ->
+        (match FD.minimal a, FD.maximal a, FD.minimal b, FD.maximal b with
+         | Some l1, Some u1, Some l2, Some u2 ->
+           (match op, ID.to_bool(FD.ne c (FD.of_const 0.)) with
+            | Le, Some true
+            | Gt, Some false -> meet_bin (FD.ending u2) (FD.starting l1)
+            | Ge, Some true
+            | Lt, Some false -> meet_bin (FD.starting l2) (FD.ending u1)
+            | Lt, Some true
+            | Ge, Some false -> meet_bin (FD.ending (Float.pred u2)) (FD.starting (Float.succ l1))
+            | Gt, Some true
+            | Le, Some false -> meet_bin (FD.starting (Float.succ l2)) (FD.ending (Float.pred u1))
+            | _, _ -> a, b)
+         | _ -> a, b)
+      | op ->
+        if M.tracing then M.tracel "inv" "Unhandled operator %a\n" d_binop op;
+        a, b
+    in
     let eval e st = eval_rv a gs st e in
     let eval_bool e st = match eval e st with `Int i -> ID.to_bool i | _ -> None in
     let set' lval v st = set a gs st (eval_lv a gs st lval) (Cilfacade.typeOfLval lval) v ~invariant:true ~ctx:(Some ctx) in
-    let rec inv_exp c exp (st:store): store =
+    let rec inv_exp c_typed exp (st:store): store =
       (* trying to improve variables in an expression so it is bottom means dead code *)
-      if ID.is_bot c then raise Deadcode;
-      match exp with
-      | UnOp (LNot, e, _) ->
+      (
+        match c_typed with
+        | `Int c -> if ID.is_bot c then raise Deadcode
+        | `Float c -> if FD.is_bot c then raise Deadcode
```

Comment:
This will never happen, since `FD.is_bot = false` Not sure if its a problem though. Also we might change this anyways if we implement the FloatDomain differently

Reply:
you are right, but with regards to future changes and in order to have a uniform interface with int and float types i would also keep it this ways

Reply:
I agree